### PR TITLE
authn.kubernetes.Resolve now behaves exactly like Kubernetes

### PR DESCRIPTION
**Pull request by dprotaso**
Prior we weren't matching host names and partial paths properly

This was breaking Gitlab and Azure tag to digest resolution in Knative (https://github.com/knative/serving/issues/12761)

### CODE REVIEW

**Comment by dprotaso** [comment]
I think I need to pull the first commit into a separate PR since I need to bump the ggcr version in the kubernetes/go.mod

Otherwise the kubernetes pkg won't pull the right ggcr min version containing the new json marshalling helpers.

**Comment by dprotaso** [comment]
Step 1: https://github.com/google/go-containerregistry/pull/1350

Will rebase this PR with go.mod updates

File: pkg/authn/kubernetes/keychain.go [hunk 1]
```diff
@@ -150,45 +157,125 @@ func NewFromPullSecrets(ctx context.Context, secrets []corev1.Secret) (authn.Key
...
+			return false, nil
 		}
-		cfg.Username = parts[0]
-		cfg.Password = parts[1]
+	}
+	// everything matches
+	return true, nil
+}
+
+func toAuthenticator(configs ...authn.AuthConfig) (authn.Authenticator, error) {
```
Comment (dprotaso) [inline pkg/authn/kubernetes/keychain.go:273]: this function doesn't need to be variadic

**Comment by imjasonh** [comment]
Should we promote this extra fuzzy URL matching to match k8s' behavior up to `pkg/authn` and use it for all Docker config handling?

Thanks so much for doing this investigation and work @dprotaso, you're a rockstar 👨‍🎤 ❤️

File: pkg/authn/kubernetes/keychain.go [hunk 2]
```diff
@@ -150,45 +157,125 @@ func NewFromPullSecrets(ctx context.Context, secrets []corev1.Secret) (authn.Key
...
-		cfg, ok = kc[key]
-		if ok {
-			break
+func (keyring *keyring) Resolve(target authn.Resource) (authn.Authenticator, error) {
+	image := target.String()
+	auths := []authn.AuthConfig{}
+
+	for _, k := range keyring.index {
+		// both k and image are schemeless URLs because even though schemes are allowed
+		// in the credential configurations, we remove them in Add.
```
[the remark this answers is not in the document]
Comment (dprotaso) [reply pkg/authn/kubernetes/keychain.go:186]: This was a comment from the k8s code - I'll update it

**Comment by dprotaso** [comment]
> Should we promote this extra fuzzy URL matching to match k8s' behavior up to pkg/authn and use it for all Docker config handling?

I was wondering if the fuzzy matching is only k8s or is that a docker config convention?

**Comment by dprotaso** [comment]
Unit test failure seems unrelated

```
--- FAIL: TestWriteIndex_Progress (2.28s)
    progress_test.go:193: saw progress revert: was high of 2659099, saw 2524410
```
https://github.com/google/go-containerregistry/runs/6024983982?check_suite_focus=true#step:4:715

**Comment by dprotaso** [comment]
presubmit looks legit - looking into that

**Comment by dprotaso** [comment]
After this I need to bump the k8schain/go.mod

**Comment by imjasonh** [comment]
> I was wondering if the fuzzy matching is only a k8s behaviour or is that a docker config convention? I would do it in a separate PR if it's not K8s convention.

Sounds like it's just K8s' version, but it seems useful to apply Postel's law here and everywhere. Maybe this means we can drop (some of?) our dep from pkg/authn -> docker/cli, and things like this:

https://github.com/google/go-containerregistry/blob/f1b729141d334006360b93a572dec46a370ea449/pkg/authn/keychain.go#L126

(not for this PR, just something to consider in the future)

**Comment by dprotaso** [comment]
> (not for this PR, just something to consider in the future)

👍

**Review by imjasonh — APPROVED** [review]

**Pull request merged**